spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java

```diff
@@ -0,0 +1,312 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.bigquery.integration.outbound;
+
+import java.io.BufferedInputStream;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.channels.Channels;
+import java.time.Duration;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.FormatOptions;
+import com.google.cloud.bigquery.Job;
+import com.google.cloud.bigquery.JobInfo.WriteDisposition;
+import com.google.cloud.bigquery.JobStatus.State;
+import com.google.cloud.bigquery.TableDataWriteChannel;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.WriteChannelConfiguration;
+
+import org.springframework.cloud.gcp.bigquery.integration.BigQuerySpringMessageHeaders;
+import org.springframework.expression.EvaluationContext;
+import org.springframework.expression.Expression;
+import org.springframework.expression.common.LiteralExpression;
+import org.springframework.integration.expression.ExpressionUtils;
+import org.springframework.integration.expression.FunctionExpression;
+import org.springframework.integration.expression.ValueExpression;
+import org.springframework.integration.handler.AbstractReplyProducingMessageHandler;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHandlingException;
+import org.springframework.util.Assert;
+import org.springframework.util.StreamUtils;
+import org.springframework.util.concurrent.SettableListenableFuture;
+
+/**
+ * A {@link org.springframework.messaging.MessageHandler} which handles sending and
+ * loading files to a BigQuery table.
+ *
+ * @author Daniel Zou
+ * @since 1.2
+ */
+public class BigQueryFileMessageHandler extends AbstractReplyProducingMessageHandler {
+
+	private final BigQuery bigQuery;
+
+	private EvaluationContext evaluationContext;
+
+	private Expression datasetNameExpression;
+
+	private Expression tableNameExpression;
+
+	private Expression formatOptionsExpression;
+
+	private boolean autoDetectSchema = true;
+
+	private WriteDisposition writeDisposition = WriteDisposition.WRITE_APPEND;
+
+	private Duration timeout = Duration.ofMinutes(5);
+
+	private Duration jobPollInterval = Duration.ofSeconds(2);
+
+	private boolean sync = false;
+
+	public BigQueryFileMessageHandler(BigQuery bigQuery) {
+		Assert.notNull(bigQuery, "BigQuery client must not be null.");
+		this.bigQuery = bigQuery;
+
+		this.datasetNameExpression =
+				new FunctionExpression<Message>(
```

Comment:
[optional] Would it be easier to read with a `headerExpression()` helper method?

Reply:
Hmm, I think i prefer the explicit style; I considered the helper method approach but preferred this to minimize a level of indirection since it is not too much extra code.

Reply:
I even prefer to have this kind of initialization at the place of property definition instead of ctor.
But that's already IMHO...